include_core/AtomicSupport.hpp

```diff
@@ -438,6 +438,80 @@ class VM_AtomicSupport
 #endif /* defined(OMR_ENV_DATA64) */
 	}
 
+	/**
+	 * Store the unsigned 32-bit value at the memory location as an atomic operation, and
+	 * return the old 32-bit value stored at the memory location.
+	 *
+	 * @param address The memory location to be updated
+	 * @param newValue The new value to be stored at memory address
+	 *
+	 * @return the value at memory location <b>address</b> BEFORE the store was attempted
+	 */
+	VMINLINE static uint32_t
+	lockExchangeU32(volatile uint32_t *address, uint32_t newValue)
+	{
+#if defined(ATOMIC_SUPPORT_STUB)
+		return 0;
+#else /* defined(ATOMIC_SUPPORT_STUB) */
+#if defined(__GNUC__)
+		return (uint32_t)__sync_lock_test_and_set(address, newValue);
+#elif defined(__xlC__) /* defined(__GNUC__) */
+		return (uint32_t)__fetch_and_swap((volatile int*)address, (int)newValue);
+#elif defined(_MSC_VER) /* defined(__GNUC__) */
+		return (uint32_t)_InterlockedExchange((volatile long *)address, (long)newValue);
+#else /* defined(__GNUC__) */
+#error "lockExchangeU32(): unsupported platform!"
+#endif /* defined(__GNUC__) */
+#endif /* defined(ATOMIC_SUPPORT_STUB) */
+	}
+
+	/**
+	 * Store the unsigned 64-bit value at the memory location as an atomic operation, and
+	 * return the old 64-bit value stored at the memory location.
+	 *
+	 * @param address The memory location to be updated
+	 * @param newValue The new value to be stored at memory address
+	 *
+	 * @return the value at memory location <b>address</b> BEFORE the store was attempted
+	 */
+	VMINLINE static uint64_t
+	lockExchangeU64(volatile uint64_t *address, uint64_t newValue)
```

Comment:
Looking at these new APIs and the assembly generated, aren't they equivalent to already existing APIs? Isn't this function a copy of `setU64` [1] effectively?

[1] https://github.com/eclipse/omr/blob/3d1baf40fa28352196d77c1a1a07b5c2f7df6145/include_core/AtomicSupport.hpp#L744-L767

Reply:
On x86, `xchg` and `cmpxchg` are different instructions. On PPC, compare-and-swap and exchange yield almost similar assembly. I think the same happens on s390x (CAS == exchange). Should we just call `setU64` from `lockExchange` on PPC and s390x?

**PPC compare-and-swap:**
```
	//__sync_synchronize
	sync
	//__sync_val_compare_and_swap
        ld 9,48(31)
        ld 8,40(31)
        ld 10,32(31)
        sync
        ldarx 7,0,9
        cmpd 7,7,8
        bne 7,.L3
        stdcx. 10,0,9
        mcrf 7,0
        bne 7,.L2
        isync
```

**PPC atomic-exchange:**
```
	//__sync_synchronize
        sync
        //__sync_lock_test_and_set
        ld 9,48(31)
        ld 10,32(31)
        ldarx 8,0,9
        stdcx. 10,0,9
        mcrf 7,0
        bne 7,.L2
        isync
```

Reply:
For Linux on Z and z/OS yes I think we can just call the `setU32` and `setU64` respectively as they will yield the exact same sequence.